cpp/ql/src/semmle/code/cpp/models/implementations/Strset.qll

```diff
@@ -0,0 +1,52 @@
+/**
+ * Provides implementation classes modeling `strset` and various similar
+ * functions. See `semmle.code.cpp.models.Models` for usage information.
+ */
+
+import semmle.code.cpp.Function
+import semmle.code.cpp.models.interfaces.ArrayFunction
+import semmle.code.cpp.models.interfaces.DataFlow
+import semmle.code.cpp.models.interfaces.Alias
+import semmle.code.cpp.models.interfaces.SideEffect
+
+/**
+ * The standard function `strset` and its assorted variants
+ */
+private class StrsetFunction extends ArrayFunction, DataFlowFunction, AliasFunction,
+  SideEffectFunction {
+  StrsetFunction() {
+    hasGlobalName([
+        "strset", "_strset", "_strset_l", "_wcsset", "_wcsset_l", "_mbsset", "_mbsset_l",
+        "_mbsnbset", "_mbsnbset_l", "_strnset", "_strnset_l", "_wcsnset", "_wcsnset_l", "_mbsnset",
+        "_mbsnset_l"
+      ])
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    // flow from the character that overrides the string
+    input.isParameter(1) and
+    (
+      output.isReturnValueDeref()
+      or
+      output.isParameterDeref(1)
```

Comment:
What is this flow for?

Reply:
This scenario (which I've also put in taint.cpp):
```cpp
void test_strset_1(char* ptr, char source) {
  _strset(ptr, source);
  sink(*ptr); // <-- That one
}
```
It doesn't work in the AST nor the IR dataflow library, though. I know why it doesn't work in IR, but I'm not sure why it doesn't for AST (I guess this is just a limitation of the AST-based approach).

Reply:
Ah, yes that's reasonable.  It should be `output.isParameterDeref(0)`.  For some reason I didn't realize you'd made a typo and was looking for a reason a parameter would taint itself. :S

Reply:
Oh! Well spotted.

Reply:
Fixed in 23eb4d2009aac295b70507f1084ef351e9195e83. That also gave us the right flow in both AST and IR.